src/Service/Api.php

```diff
@@ -292,6 +311,63 @@ private function getConnectorOptions() {
         return $connectorOptions;
     }
 
+    /**
+     * Waits for a lock, if it exists, or creates one.
+     *
+     * @param string $lockName
+     *   A unique name for the lock.
+     * @param callable|null $onWait
+     *   A function to run when waiting starts.
+     * @param callable|null $check
+     *   A function to run each time the interval has passed. If it returns a
+     *   non-null value, waiting will stop, and the value will be returned
+     *   from this method.
+     * @param int $intervalMs
+     *   A waiting interval in milliseconds.
+     * @param int $timeLimit
+     *   A time limit in seconds.
+     *
+     * @return mixed|null
+     */
+    public function lock($lockName, callable $onWait = null, callable $check = null, $intervalMs = 300, $timeLimit = 15)
+    {
+        $runOnWait = false;
+        $cacheKey = 'lock:' . $lockName;
+        $start = \time();
+        while (\time() - $start < $timeLimit) {
+            $cached = $this->cache->fetch($cacheKey);
```

Comment:
Unfortunately, although it's convenient, this should probably not use the cache because that can be disabled in config.